### PR TITLE
Allow onfail req to be used with multiple reqs in a single state

### LINKED ISSUE #59026: Multiple requisites not applied in 3002
Hi, 

We are upgrading Salt from 3000.3 to 3002 and are experiencing unusual behavior under multiple requisites.

Given the following state:
```
A:
  cmd.run:
    - name: echo Hi A

B:
  cmd.run:
    - name: echo Hi B
    - require:
      - cmd: A

C:
  cmd.run:
    - name: echo Hi C
    - onfail:
      - cmd: B
    - onchanges:
      - cmd: A
```
We would expect step C to be executed only if B fails and A successfully executes with changes
That is exactly what we see in 3000.3 but not in 3002

**Salt 3002**
```
local:
----------
          ID: A
    Function: cmd.run
        Name: echo Hi A
      Result: True
     Comment: Command "echo Hi A" run
     Started: 13:44:24.784490
    Duration: 10.008 ms
     Changes:
              ----------
              pid:
                  31900
              retcode:
                  0
              stderr:
              stdout:
                  Hi A
----------
          ID: B
    Function: cmd.run
        Name: echo Hi B
      Result: True
     Comment: Command "echo Hi B" run
     Started: 13:44:24.795096
    Duration: 8.29 ms
     Changes:
              ----------
              pid:
                  31901
              retcode:
                  0
              stderr:
              stdout:
                  Hi B
----------
          ID: C
    Function: cmd.run
        Name: echo Hi C
      Result: True
     Comment: Command "echo Hi C" run
     Started: 13:44:24.804015
    Duration: 7.198 ms
     Changes:
              ----------
              pid:
                  31902
              retcode:
                  0
              stderr:
              stdout:
                  Hi C

Summary for local
------------
Succeeded: 3 (changed=3)
Failed:    0
------------
Total states run:     3
Total run time:  25.496 ms
```

**Salt 3000.3**
```
local:
----------
          ID: A
    Function: cmd.run
        Name: echo Hi A
      Result: True
     Comment: Command "echo Hi A" run
     Started: 13:53:13.793835
    Duration: 11.943 ms
     Changes:
              ----------
              pid:
                  12954
              retcode:
                  0
              stderr:
              stdout:
                  Hi A
----------
          ID: B
    Function: cmd.run
        Name: echo Hi B
      Result: True
     Comment: Command "echo Hi B" run
     Started: 13:53:13.806508
    Duration: 9.238 ms
     Changes:
              ----------
              pid:
                  12955
              retcode:
                  0
              stderr:
              stdout:
                  Hi B
----------
          ID: C
    Function: cmd.run
        Name: echo Hi C
      Result: True
     Comment: State was not run because onfail req did not change
     Started: 13:53:13.816423
    Duration: 0.016 ms
     Changes:

Summary for local
------------
Succeeded: 3 (changed=2)
Failed:    0
------------
Total states run:     3
Total run time:  21.197 ms
```

Here the result of `salt --versions-report`
```
Salt Version:
           Salt: 3002

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.8.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.8.1
        libgit2: Not Installed
       M2Crypto: 0.35.2
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: Not Installed
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Nov 16 2020, 16:55:22)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 17.0.0
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.4

System Versions:
           dist: centos 7 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-1062.9.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7 Core

```

```
Salt Version:
           Salt: 3000.3

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 2.8.1
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.8.1
        libgit2: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.8 (default, Nov 16 2020, 16:55:22)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 17.0.0
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.1.6

System Versions:
           dist: centos 7.9.2009 Core
         locale: ANSI_X3.4-1968
        machine: x86_64
        release: 4.19.128-microsoft-standard
         system: Linux
        version: CentOS Linux 7.9.2009 Core

```

### PR DESCRIPTION
**Pull request by xeacott**
### What does this PR do?
Allows users to have `onfail` used with other requisites

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/59026

* Other potential fix:
```
@@ -2796,6 +2800,8 @@ class State:
             status = "onfail"
         elif "onchanges" in fun_stats and "onchangesmet" not in fun_stats:
             status = "onchanges"
+        elif "onfail" in fun_stats and "onchangesmet" in fun_stats and "met" not in fun_stats:
+            status = "onfail"
         elif "change" in fun_stats:
             status = "change"
         else:
```

### Previous Behavior
Remove this section if not relevant

### New Behavior
Remove this section if not relevant

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Review by garethgreenaway — APPROVED** [review]

**Comment by garethgreenaway** [comment]
@xeacott Needs a changelog file and since this is changing the behavior of the state engine (fixing broken functionality) we should include a note about the change in the release notes.

**Review by garethgreenaway — CHANGES_REQUESTED** [review]
Needs changelog and entry in release notes.

**Review by garethgreenaway — DISMISSED** [review]

**Review by frogunder — DISMISSED** [review]

File: changelog/59026.fixed [hunk 1]
```diff
@@ -0,0 +1 @@
+Add extra onfail req check in the state engine to allow onfail to be used with onchanges and other reqs in the same state
```
Comment (Ch3LL) [inline changelog/59026.fixed:1]: If this is changing behavior we should also name the changelog file 59026.changed

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: changelog/59026.fixed (hunk 1 above)
Comment (xeacott) [reply changelog/59026.fixed:1]: Agreed

**Review by Ch3LL — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Review by garethgreenaway — APPROVED** [review]

**Pull request merged**